### ISSUE TITLE
`export` keyword not transpiled properly

### ISSUE THREAD

**Issue by gnh1201** [swc-project/swc#8365]
### Describe the bug

When attempting to transpile TypeScript to ES3 or ES5, the export keyword does not transpile properly.

### Input code

```typescript
let message: string = 'Hello, World!';

function main(args) {
    console.log(message);
}

export {main}
```

### Config

```json
// es3
{
    "jsc": {
        "target": "es3"
    }
}

// core-js 3.26.1 (my real world)
{
    "jsc": {
        "parser": {
            "syntax": "ecmascript",
            "jsx": false
        }
    },
    "env": {
        "mode": "usage",
        "coreJs": "3.26.1"
    }
}
```

### Playground link (or link to the minimal reproduction)

https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAAy3MQQqDMBBG4f2c4u%2FKCKUHqHTvDVwHnYbAZKZkIgji3VXo239PuKGwe0z8hreaNeGDbmQRe2KyKsujG4i%2Bq84tm6LErCHW5D12wtVs6ib8Ekvhf%2BoHOoh4%2B1lt2G9xnMR%2FNN1oAAAA&config=H4sIAAAAAAAAA1WPSw7DIAwF9zkF8rrbdtE79BCIOhERP9mOVBTl7oUE0maH3xszsA5KwcwGnmotxzIkTYx0ziXhHER%2FSgKSE7IhmwRuvRWu1agd4x5tRwOiaUKpW8j3hoOLkbHjLfM22DH%2FC030iZD5ClZUh8nhVTc0Jfj4XvayfaQ%2B9tA%2F4Ad12XkxWH71TaEFh%2B0LYuVI0xQBAAA%3D

### Expected behavior

(using Typescript 4.9.4 from Microsoft)

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
var message = 'Hello, World!';
function main(args) {
    console.log(message);
}
exports.main = main;
```

### Actual behavior

(using SWC v1.3.101-nightly-20231201.1)

```js
var message = "Hello, World!";
function main(args) {
    console.log(message);
}
export { main };
```

### Version

v1.3.101-nightly-20231201.1

**Comment by kdy1** [comment]
Your module config is wrong. https://swc.rs/docs/configuration/modules

**Comment by gnh1201** [comment]
I've tried all available module types in the playground, but it still doesn't transpile `export { main };` and displays it as is. I've modified the .swcrc, but it remains the same. The expected transpilation result should be `exports.main = main;`.

**Comment by kdy1** [comment]
https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAAy3MQQqDMBBG4f2c4u%2FKCKUHqHTvDVwHnYbAZKZkIgji3VXo239PuKGwe0z8hreaNeGDbmQRe2KyKsujG4i%2Bq84tm6LErCHW5D12wtVs6ib8Ekvhf%2BoHOoh4%2B1lt2G9xnMR%2FNN1oAAAA&config=H4sIAAAAAAAAA1WPSw7DMAhE9z6FxbrbbnqHHsJySeTIPwGRakW5e%2B3UTpsdzAw82JTWsLCFh95qWZtsiJHOvipcoph3VUBKRrbkssBtuMLNmoxnPKT964AYmlHaFPK9x8GnxDjiXQsuuqn8A20KmZD5GmxRE2ePV5zqSAjptR5mf6Qd2%2FB1W0hxYfglB%2FHcDo6fY1xoRbV%2FALu7P%2BkZAQAA

**Comment by gnh1201** [comment]
I apologize. I'm aware that there is an issue related to accessors in the engine I'm using. Microsoft's TypeScript transpiler likely automatically recognizes that specification.

SWC provided results tailored for engines that support accessors (defineProperty + getter + setter), and this is the correct outcome. It seems I momentarily misunderstood and made a mistake.

I will document this issue separately and reach out again later for further inquiry.

**Comment by gnh1201** [comment]
Is it possible to avoid the accessor pattern, which is one of the patterns involving the use of defineProperty?

**Comment by kdy1** [comment]
Yes, you can write a Wasm plugin if you want.

`@swc/core` will never provide such an option as ESM being immutable is part of the specification.

These links will help you implement Wasm plugins.

https://github.com/swc-project/swc/pull/4758 is the PR for implementing immutability for ESMs, and the old transform code lives at https://github.com/swc-project/swc/blob/f41c82ac67717afb7f5deef5ed306e3d2b4a07dd/crates/swc_ecma_transforms_module/src/common_js.rs


You can just copy old commonjs code, build it as a Wasm plugin, and use it.

**Comment by gnh1201** [comment]
Understood.

However, while it's correct to use the accessor pattern to comply with the ESM spec in es5, using the accessor pattern in an engine at the es3 level can cause parse errors, posing a problem.

In practice, the result including the accessor pattern becomes code that cannot be used in es3. I would like to inquire whether strictly adhering to the ESM spec is preferable over the actual usage environment.

**Comment by kdy1** [comment]
Our main PoV is that something not following the specification can be implemented as a non-official userland plugin.

**Comment by gnh1201** [comment]
Since there are no issues with Babel and Microsoft's transpiler, I currently don't understand the explanation that there might be a problem due to my engine not adhering to the spec.

**Comment by magic-akari** [comment]
Yes, this is a potential issue.

## Transpiling from ES5 to ES3

> [!NOTE]  
Downgrading from ES5 to ES3 should not be considered to converte ES Modules to CommonJS.

While ES Modules were introduced after ES5, in practice, we handle ES syntax and ES Modules separately.
When downgrading from ES5 to ES3, the ES Module syntax (import/export) will be preserved.
In fact, we have handled several issues related to ES3 reserved words and import/export, which have practical applications.

In typical use cases, swc is used to downgrade to ES3, while other bundling tools like Rollup handle the packaging. In the final output, ES Module syntax will be eliminated.

## Why we chose getters for transpiling to CommonJS

The main reason is that using getters allows us to achieve ES Module live binding with minimal and optimized code. Its immutability coincidentally aligns with the ES spec.

CommonJS is not part of the ES standard. When deciding to transpile to CommonJS, we assume everything is related to the node runtime. 

By referencing the documentation, [`Object.defineProperty`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty) and [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) are available in node 0.10.0, which is nearly the first version of node. 

Therefore, we believe the current implementation is sufficiently compatible with the node runtime.

**Comment by gnh1201** [comment]
@magic-akari Thank you for the precise explanation.

Based on my testing, the module type in SWC that does not exhibit the accessor pattern is recognized as SystemJS. I will need to test soon whether I can apply this in my environment.

Thank you.
